File: backend/apps/trak/models/address.py
from django.db import models

from lib.rcrainfo import lookups as lu


class Address(models.Model):
    street_number = models.CharField(
        max_length=12,
        null=True,
        blank=True,
    )
    address1 = models.CharField(
        verbose_name='Address 1',
        max_length=50,
    )
    address2 = models.CharField(
        verbose_name='Address 2',
        max_length=50,
        default=None,
        null=True,
        blank=True,
    )
    city = models.CharField(
        max_length=25,
    )
    state = models.CharField(
        max_length=3,
        null=True,
        blank=True,
        choices=lu.STATES,
    )
    country = models.CharField(
        max_length=3,
        null=True,
        blank=True,
        choices=lu.COUNTRIES,
    )
    zip = models.CharField(
        null=True,
        blank=True,
        max_length=5,
    )

    def __str__(self):
        if self.street_number:
            return f'{self.street_number} {self.address1}'
        else:
            return f' {self.address1}'

File: client/src/types/site/address.ts
import { z } from 'zod';

/**
 * Locality Schema for EPA's Rcrainfo which contains information on a geographic region (state, country)
 * used for RCRA sites and handlers on a hazardous waste manifest
 */
// ToDo: rcraLocality code and name should be validated togather so codes correspond geographic regions
//  For example: if code === "TX" is true then name === "Texas" should also be true.
export const rcraLocalitySchema = z.object({
  code: z.string(),
  name: z.string().optional(),
});

/**
 * Address Schema  information for handlers on a hazardous waste manifest
 */
export const rcraAddressSchema = z.object({
  address1: z.string(),
  address2: z.string().optional(),
  city: z.string().optional(),
  country: rcraLocalitySchema.optional(),
  state: rcraLocalitySchema,
  streetNumber: z.string().optional(),
  zip: z.string().optional(),
});

/**
 * format of address information stored by EPA's RCRAInfo
 */
export type RcraAddress = z.infer<typeof rcraAddressSchema>;

/**
 * format of regional data including a code (e.g., TX) and name (e.g., Texas)
 * stored by EPA's RCRAInfo
 */
export type RcraLocality = z.infer<typeof rcraLocalitySchema>;
